res/res_pjsip/config_transport.c

```diff
@@ -828,17 +828,50 @@ static int transport_apply(const struct ast_sorcery *sorcery, void *obj)
 	} else if (transport->type == AST_TRANSPORT_TCP) {
 		pjsip_tcp_transport_cfg cfg;
 		static int option = 1;
+		int sockopt_count = 0;
 
 		pjsip_tcp_transport_cfg_default(&cfg, temp_state->state->host.addr.sa_family);
 		cfg.bind_addr = temp_state->state->host;
 		cfg.async_cnt = transport->async_operations;
 		set_qos(transport, &cfg.qos_params);
+
 		/* sockopt_params.options is copied to each newly connected socket */
-		cfg.sockopt_params.options[0].level = pj_SOL_TCP();
-		cfg.sockopt_params.options[0].optname = pj_TCP_NODELAY();
-		cfg.sockopt_params.options[0].optval = &option;
-		cfg.sockopt_params.options[0].optlen = sizeof(option);
-		cfg.sockopt_params.cnt = 1;
+		cfg.sockopt_params.options[sockopt_count].level = pj_SOL_TCP();
+		cfg.sockopt_params.options[sockopt_count].optname = pj_TCP_NODELAY();
+		cfg.sockopt_params.options[sockopt_count].optval = &option;
+		cfg.sockopt_params.options[sockopt_count].optlen = sizeof(option);
+		sockopt_count++;
+
+		if (transport->tcp_keepalive_enable) {
```

Comment:
What's the best way to take this into account @jcolp?

Is a compile-time check a la #if PJ_MAX_SOCKOPT_PARAMS >= 5 acceptable here or do you want something more to be done at runtime? And should we kill off in its entirety if we cannot apply all settings? The default limit is 4, so we could lose the least important setting, but that's also probably not intuitive to users.

Reply:
We use a compile time check for other things so that is fine. I think if the full set isn't possible, then they should all be disabled, since a user may want the one left out which leads down a rabbit hole of needing to make that configurable. It also has to be documented as such, and at runtime a message output about it.